OpenRA.Mods.Cnc/FileSystem/MegFile.cs

```diff
@@ -0,0 +1,206 @@
+			/// </summary>
+			/// <param name="reader">The reader of the stream</param>
+			private void ParseMegHeader(BinaryReader reader)
+			{
+				var dataStartOffset = reader.ReadUInt32();
+				var numFileNames = reader.ReadUInt32();
+				var numFiles = reader.ReadUInt32();
+				var fileNameTableSize = reader.ReadUInt32();
+
+				// the file names are an indexed array of strings
+				for (uint i = 0; i < numFileNames; i++)
+				{
+					var fileNameLength = reader.ReadUInt16();
+					var fileNameBytes = reader.ReadBytes(fileNameLength);
+					var fileName = Encoding.ASCII.GetString(fileNameBytes);
+
+					contents.Add(fileName);
+				}
+
+				// the header indicates where we should be, so verify it
+				if (reader.BaseStream.Position != fileNameTableSize + HeaderFileNameTableStartOffset)
+					throw new Exception("File name table in .meg file inconsistent");
+
+				// now we load each file entry and associated info
+				for (var i = 0; i < numFiles; i++)
+				{
+					var flags = reader.ReadUInt16();
+					if (flags != 0) throw new Exception("Encrypted files are not supported or expected.");
+
+					var crc = reader.ReadUInt32();
+					var fileRecordIndex = reader.ReadUInt32();
+					var fileSize = reader.ReadUInt32();
+					var fileStartOffset = reader.ReadUInt32();
+					var fileNameIndex = reader.ReadUInt16();
+
+					fileData.Add(new MegFileContentReference(crc, fileRecordIndex, fileSize, fileStartOffset, fileNameIndex));
+				}
+
+				if (reader.BaseStream.Position != dataStartOffset)
+				{
+					throw new Exception("Expected to be at data start offset");
+				}
+			}
+
+			public string Name { get { return name; } }
+
+			public IEnumerable<string> Contents { get { return contents; } }
+
+			public IReadOnlyDictionary<string, PackageEntry> Index
+			{
+				get
+				{
+					var absoluteIndex = Contents.ToDictionary(e => e, e =>
+					{
+						var index = contents.IndexOf(e);
+						var reference = fileData.FirstOrDefault(a => a.FileNameTableRecordIndex == index);
+						return new PackageEntry(reference.Crc32, reference.FileStartOffsetInBytes, reference.FileSizeInBytes);
+					});
+					return new ReadOnlyDictionary<string, PackageEntry>(absoluteIndex);
+				}
+			}
+
+			public bool Contains(string filename)
+			{
+				return contents.Contains(filename);
+			}
+
+			public void Dispose()
+			{
+				s.Dispose();
+			}
+
+			public Stream GetStream(string filename)
+			{
+				// look up the index of the filename
+				var index = contents.IndexOf(filename);
+				var reference = fileData.FirstOrDefault(a => a.FileNameTableRecordIndex == index);
+
+				return SegmentStream.CreateWithoutOwningStream(s, reference.FileStartOffsetInBytes, (int)reference.FileSizeInBytes);
+			}
+
+			public IReadOnlyPackage OpenPackage(string filename, OpenRA.FileSystem.FileSystem context)
+			{
+				IReadOnlyPackage package;
+				var childStream = GetStream(filename);
```

Comment:
Maybe use a `using` block instead of explicitly disposing? Or do we rely on this not being disposed if `TryParsePackage` returns true?

Reply:
I copied this logic from MixFile - I am unsure if it is correct here or not. The way I understand this code is if the .meg file contains a zip file, that this method would be called to have the .zip contents be able to be extracted through a Zip loader. I am assuming what's applies for MixFile does here too, but happy to be corrected.